test/jdk/java/nio/channels/FileChannel/Transfer2GPlus.java

```diff
@@ -83,22 +83,28 @@ private static void testToFileChannel(Path src, byte[] expected)
         try (FileChannel srcCh = FileChannel.open(src)) {
             try (FileChannel dstCh = FileChannel.open(dst,
                  StandardOpenOption.READ, StandardOpenOption.WRITE)) {
-                long n;
-                if ((n = srcCh.transferTo(0, LENGTH, dstCh)) < LENGTH) {
+                long total = 0L;
+                if ((total = srcCh.transferTo(0, LENGTH, dstCh)) < LENGTH) {
                     if (!Platform.isLinux())
-                        throw new RuntimeException("Transfer too small: " + n);
+                        throw new RuntimeException("Transfer too small: " + total);
 
-                    if (n != 0x7ffff000)
-                        throw new RuntimeException("Unexpected transfer size: " + n);
-                    if ((n += srcCh.transferTo(n, LENGTH, dstCh)) != LENGTH)
-                        throw new RuntimeException("Unexpected total size: " + n);
+                    // if this point is reached we're on Linux
+                    if (total > LINUX_MAX_TRANSFER_SIZE)
+                        throw new RuntimeException("Unexpected transfer size: " + total);
+
+                    do {
+                        long n = srcCh.transferTo(total, LENGTH, dstCh);
+                        if (n == 0)
+                            break;
+                        total += n;
+                    } while (total < LENGTH);
```

Comment:
Could it be possible where the `total` is larger than `LINUX_MAX_TRANSFER_SIZE` here after the loop? If so, the test below only checks whether the `dstCh` is smaller than `LENGTH` and proceeds without the error reported.

Reply:
On Linux the `transferTo()` method uses `sendfile()` for the code in question. The specification of [sendfile()](https://man7.org/linux/man-pages/man2/sendfile.2.html) states that it will transfer at most `0x7ffff000` (`LINUX_MAX_TRANSFER_SIZE`) bytes. So here we are only checking for an aberration which is not essential to the test. The important content is when `transferTo()` is supposed to actually copy the full length in one shot. This Linux part of the test is just a drive-by sanity check.

Reply:
Thanks for the explanation. Then I would think adding the same check in the loop for consistency, i.e., check the transferred size `n` in one shot is no more than `LINUX_MAX_TRANSFER_SIZE`  after line 98.

Reply:
I can do that. Thanks.